File: windows/Services/MsuPackImporter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Text.Json;
using LTTPRandomizerGenerator.Models;

namespace LTTPRandomizerGenerator.Services;

public record MsuBundleExportResult(int TracksWritten, int TracksSkipped);

public static class MsuPackImporter
{
    private static readonly string LibraryFolder = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "LTTPRandomizerGenerator", "MsuLibrary");

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNameCaseInsensitive = true
    };

    public static (MsuPlaylist? playlist, string? error) Import(string zipPath)
    {
        try
        {
            using var zip = ZipFile.OpenRead(zipPath);

            var manifestEntry = zip.GetEntry("manifest.json");
            if (manifestEntry is null)
                return (null, "Invalid pack: manifest.json not found.");

            PackManifest manifest;
            using (var ms = manifestEntry.Open())
            using (var reader = new StreamReader(ms))
            {
                var json = reader.ReadToEnd();
                var parsed = JsonSerializer.Deserialize<PackManifest>(json, JsonOptions);
                if (parsed is null)
                    return (null, "Invalid pack: manifest.json could not be read.");
                manifest = parsed;
            }

            if (manifest.Version != 1)
                return (null, $"Unsupported pack version: {manifest.Version}");

            if (string.IsNullOrWhiteSpace(manifest.Name))
                manifest.Name = Path.GetFileNameWithoutExtension(zipPath);

            string safePackName = string.Concat(manifest.Name.Split(Path.GetInvalidFileNameChars())).Trim();
            if (string.IsNullOrEmpty(safePackName))
                safePackName = "imported-pack";

            string destFolder = Path.Combine(LibraryFolder, "Imported", safePackName);
            Directory.CreateDirectory(destFolder);

            var tracks = new Dictionary<string, string>();

            foreach (var (slotKey, entryName) in manifest.Tracks)
            {
                var zipEntry = zip.GetEntry(entryName);
                if (zipEntry is null) continue;

                string destFileName = Path.GetFileName(entryName);
                string destPath = Path.Combine(destFolder, destFileName);

                // Guard against path traversal
                string fullDest = Path.GetFullPath(destPath);
                string fullFolder = Path.GetFullPath(destFolder) + Path.DirectorySeparatorChar;
                if (!fullDest.StartsWith(fullFolder, StringComparison.OrdinalIgnoreCase))
                    continue;

                if (!File.Exists(destPath))
                {
                    using var src = zipEntry.Open();
                    using var dst = File.Create(destPath);
                    src.CopyTo(dst);
                }

                tracks[slotKey] = destPath;
            }

            var playlist = new MsuPlaylist
            {
                Name = manifest.Name,
                Tracks = tracks
            };

            return (playlist, null);
        }
        catch (Exception ex)
        {
            return (null, $"Import failed: {ex.Message}");
        }
    }

    public static (MsuBundleExportResult? result, string? error) Export(string destZipPath, MsuPlaylist playlist)
    {
        int written = 0;
        int skipped = 0;

        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(destZipPath)!);

            using var zip = ZipFile.Open(destZipPath, ZipArchiveMode.Create);
            var manifestTracks = new Dictionary<string, string>();

            foreach (var (slotKey, pcmPath) in playlist.Tracks)
            {
                if (!File.Exists(pcmPath)) { skipped++; continue; }
                if (!int.TryParse(slotKey, out int slot)) { skipped++; continue; }

                string entryName = $"tracks/{slot:D2}.pcm";
                var entry = zip.CreateEntry(entryName, CompressionLevel.NoCompression);
                using (var entryStream = entry.Open())
                using (var fileStream = File.OpenRead(pcmPath))
                    fileStream.CopyTo(entryStream);

                manifestTracks[slotKey] = entryName;
                written++;
            }

            var manifest = new PackManifest
            {
                Version = 1,
                Name = playlist.Name,
                Tracks = manifestTracks
            };

            var manifestJson = JsonSerializer.Serialize(manifest, JsonOptions);
            var manifestEntry = zip.CreateEntry("manifest.json", CompressionLevel.Optimal);
            using (var manifestStream = manifestEntry.Open())
            using (var writer = new StreamWriter(manifestStream))
                writer.Write(manifestJson);

            return (new MsuBundleExportResult(written, skipped), null);
        }
        catch (Exception ex)
        {
            try { if (File.Exists(destZipPath)) File.Delete(destZipPath); } catch { }
            return (null, $"Export failed: {ex.Message}");
        }
    }

    private class PackManifest
    {
        public int Version { get; set; } = 1;
        public string Name { get; set; } = string.Empty;
        public Dictionary<string, string> Tracks { get; set; } = new();
    }
}

File: android/app/src/main/java/com/lttprandomizer/MsuPackImporter.kt
package com.lttprandomizer

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File
import java.io.InputStream
import java.io.OutputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

object MsuPackImporter {
    private val json = Json { prettyPrint = true; ignoreUnknownKeys = true }

    @Serializable
    private data class PackManifest(
        val version: Int = 1,
        val name: String = "",
        val tracks: Map<String, String> = emptyMap()
    )

    fun import(context: Context, inputStream: InputStream): Pair<MsuPlaylist?, String?> {
        return try {
            val zipStream = ZipInputStream(inputStream)
            var manifest: PackManifest? = null
            val extractedFiles = mutableMapOf<String, ByteArray>()

            var entry: ZipEntry? = zipStream.nextEntry
            while (entry != null) {
                val name = entry.name
                if (name == "manifest.json") {
                    val text = zipStream.bufferedReader().use { it.readText() }
                    manifest = json.decodeFromString<PackManifest>(text)
                } else if (!entry.isDirectory) {
                    extractedFiles[name] = zipStream.readBytes()
                }
                zipStream.closeEntry()
                entry = zipStream.nextEntry
            }
            zipStream.close()

            if (manifest == null)
                return Pair(null, "Invalid pack: manifest.json not found.")

            if (manifest.version != 1)
                return Pair(null, "Unsupported pack version: ${manifest.version}")

            val packName = manifest.name.ifBlank { "imported-pack" }
            val safeName = packName.replace(Regex("[^a-zA-Z0-9._\\- ]"), "").trim().ifEmpty { "imported-pack" }

            val destDir = File(context.filesDir, "msu_library/Imported/$safeName")
            destDir.mkdirs()

            val tracks = mutableMapOf<String, String>()

            for ((slotKey, entryName) in manifest.tracks) {
                val data = extractedFiles[entryName] ?: continue
                val destFileName = entryName.substringAfterLast('/')
                val destFile = File(destDir, destFileName)

                // Guard against path traversal
                if (!destFile.canonicalPath.startsWith(destDir.canonicalPath + File.separator))
                    continue

                if (!destFile.exists()) {
                    destFile.writeBytes(data)
                }

                tracks[slotKey] = destFile.absolutePath
            }

            Pair(MsuPlaylist(name = packName, tracks = tracks), null)
        } catch (ex: Exception) {
            Pair(null, "Import failed: ${ex.message}")
        }
    }

    fun export(outputStream: OutputStream, playlist: MsuPlaylist): String? {
        return try {
            var written = 0
            val zipOut = ZipOutputStream(outputStream)
            val manifestTracks = mutableMapOf<String, String>()

            for ((slotKey, pcmPath) in playlist.tracks) {
                val file = File(pcmPath)
                if (!file.exists()) continue
                val slot = slotKey.toIntOrNull() ?: continue

                val entryName = "tracks/${slot.toString().padStart(2, '0')}.pcm"
                zipOut.putNextEntry(ZipEntry(entryName))
                file.inputStream().use { it.copyTo(zipOut) }
                zipOut.closeEntry()

                manifestTracks[slotKey] = entryName
                written++
            }

            val manifest = PackManifest(version = 1, name = playlist.name, tracks = manifestTracks)
            val manifestJson = json.encodeToString(manifest)
            zipOut.putNextEntry(ZipEntry("manifest.json"))
            zipOut.write(manifestJson.toByteArray())
            zipOut.closeEntry()

            zipOut.close()
            null
        } catch (ex: Exception) {
            "Export failed: ${ex.message}"
        }
    }
}
